app/assets/javascripts/map.js

```diff
@@ -125,25 +207,210 @@ class Map {
+    coordArray.push(firstCoord);
+    return coordArray.join("|");
+  }
+
+  handleExpeditionInfoClick(event) {
+    Dom.showNode(document.getElementById('summaryExpeditionOptions'));
+    this.serviceExpeditionCost = Number.parseFloat(event.target.value);
+    this.serviceExpeditionLabel = event.target.dataset.label;
+    this.serviceExpeditionTime = event.target.id;
+    document.getElementById('summaryServiceExpeditionTime').innerText = this.serviceExpeditionLabel
+    this.updateAmount('serviceExpeditionDue', this.serviceExpeditionCost, 2)
+    this.updateGrandTotal();
+  }
+
+  handleAddBagClick() {
+    this.saltBagsQuantity += 1;
+    this.updateSaltBagsTotals();
   }
+
+  handleRemoveBagClick() {
+    if (this.saltBagsQuantity > 0) {
+      this.saltBagsQuantity -= 1;
+      this.updateSaltBagsTotals();
+    }
+  }
+  
+  updateSaltBagsTotals() {
+    this.saltBagsDue = this.saltBagsQuantity * this.saltBagPrice;
+    this.updateTotalSaltBags();
+    this.updateGrandTotal();
+  }
+  
+  updateTotalSaltBags() {
+    this.updateAmount('numberOfBags', this.saltBagsQuantity, 0);
+    Dom.showNode(document.getElementById("summarySaltBags"));
+    this.updateAmount('summaryNumberOfBags', this.saltBagsQuantity, 0);
+    this.updateAmount('saltBagsDue', this.saltBagsDue, 2);
+  }
+
+  handleDoneAddOns() {
+    this.showNextSection('collapseAddOns', 'collapseContactAndPaymentInfo');
+    document.getElementById('checkMarkAddOns').classList.remove('hidden');
+
+  }
+
+  updateGrandTotal() {
+    let grandTotalBeforeDiscount = this.subTotal + this.serviceExpeditionCost + this.saltBagsDue;
+    this.updateDiscount(grandTotalBeforeDiscount);
+    this.grandTotal = Math.max(grandTotalBeforeDiscount, priceList.MIN_CHARGE) * this.discount;
+    if (this.grandTotal > priceList.MIN_CHARGE) {
+      Dom.hideNode(document.getElementById('minChargeNote'))
+    } else {
+      Dom.showNode(document.getElementById('minChargeNote'))
+    }
+    this.updateAmount('totalDue', this.grandTotal, 2)
+  }
+
+  updateDiscount(grandTotalBeforeDiscount) {
+    this.updateAmount('summaryDiscountAmount', this.discount * 100, 0)
+    this.updateAmount('totalDiscount', grandTotalBeforeDiscount * (- this.discount), 2)
+  }
+
+  onCreditCardInputChange(cardInputIsComplete) {
+    this.isCreditCardComplete = cardInputIsComplete;
+    this.enableSubmitButton();
+    Dom.hideNode(document.getElementById("contact_info_errors"))
+  }
+
+  enableSubmitButton() {
+    const requiredFieldsAreNotEmpty = ['quote_email', 'quote_phone_number'].every((id) => {
+      return document.getElementById(id).value !== "";
+    });
+
+    if (requiredFieldsAreNotEmpty && this.isCreditCardComplete) {
+      document.getElementById('submitPayment').classList.remove('disabled');
+    }
+  }
+
+  handlePaymentSubmit(event) {
+    event.preventDefault();
+    document.getElementById('submitPayment').classList.add('disabled');
+    this.stripe.getToken(this.handleTokenReceived);
```

Comment:
If there are card validation errors from Stripe, is the button re-enabled? You might have to pass `this.stripe.getToken` two callbacks: one for success and one for failure (which would then let you move the error handling _out_ of `stripe.js` which is currently fiddling with the error display in there, where it maybe doesn't belong)